tsgo panics when destructuring pattern has no leaf pattern and targeting ES2017

## Stack trace

```
panic: runtime error: index out of range [-1] [recovered, repanicked]

goroutine 103 [running]:
sync.(*WaitGroup).Go.func1.1()
        sync/waitgroup.go:251 +0x48
panic({0x106003d80?, 0x436900040180?})
        runtime/panic.go:860 +0x12c
github.com/microsoft/typescript-go/internal/transformers.(*flattener).flattenDestructuringBinding(0x4369008193b0, 0x436900116340?, 0x0?, 0x0?)
        github.com/microsoft/typescript-go/internal/transformers/destructuring.go:261 +0x5d4
github.com/microsoft/typescript-go/internal/transformers.FlattenDestructuringBinding(0x436901a9a780, 0x436900116340, 0x0, 0x1, 0x1, 0x0)
        github.com/microsoft/typescript-go/internal/transformers/destructuring.go:72 +0xac
github.com/microsoft/typescript-go/internal/transformers/estransforms.(*objectRestSpreadTransformer).visitVariableDeclarationWorker(0x436901a9a780, 0x436900116340, 0x1)
        github.com/microsoft/typescript-go/internal/transformers/estransforms/objectrestspread.go:436 +0x80
github.com/microsoft/typescript-go/internal/transformers/estransforms.(*objectRestSpreadTransformer).visitVariableDeclaration(0x436901a9a780, 0x0?)
        github.com/microsoft/typescript-go/internal/transformers/estransforms/objectrestspread.go:426 +0x30
github.com/microsoft/typescript-go/internal/transformers/estransforms.(*objectRestSpreadTransformer).visit(0x436901a9a780, 0x436900116340)
        github.com/microsoft/typescript-go/internal/transformers/estransforms/objectrestspread.go:47 +0x440
github.com/microsoft/typescript-go/internal/ast.(*NodeVisitor).VisitSlice(0x43690068fd40, {0x4369001380e0, 0x1, 0x1})
        github.com/microsoft/typescript-go/internal/ast/visitor.go:148 +0xa0
github.com/microsoft/typescript-go/internal/ast.(*NodeVisitor).VisitNodes(0x43690068fd40, 0x436900152080)
        github.com/microsoft/typescript-go/internal/ast/visitor.go:99 +0x3c
github.com/microsoft/typescript-go/internal/ast.(*NodeVisitor).visitNodes(0x436901837778?, 0x1051c304c?)
        github.com/microsoft/typescript-go/internal/ast/visitor.go:236 +0x44
github.com/microsoft/typescript-go/internal/ast.(*VariableDeclarationList).VisitEachChild(0x436900152100, 0x43690068fd40)
        github.com/microsoft/typescript-go/internal/ast/ast_generated.go:1942 +0x30
github.com/microsoft/typescript-go/internal/ast.(*Node).VisitEachChild(...)
        github.com/microsoft/typescript-go/internal/ast/ast.go:197
github.com/microsoft/typescript-go/internal/ast.(*NodeVisitor).VisitEachChild(...)
        github.com/microsoft/typescript-go/internal/ast/visitor.go:191
github.com/microsoft/typescript-go/internal/transformers/estransforms.(*objectRestSpreadTransformer).visit(0x436901a9a780, 0x436900152100)
        github.com/microsoft/typescript-go/internal/transformers/estransforms/objectrestspread.go:67 +0x530
github.com/microsoft/typescript-go/internal/ast.(*NodeVisitor).VisitNode(0x43690068fd40?, 0x43690015e000?)
        github.com/microsoft/typescript-go/internal/ast/visitor.go:51 +0x30
github.com/microsoft/typescript-go/internal/ast.(*NodeVisitor).visitNode(0x43690071a0c0?, 0x0?)
        github.com/microsoft/typescript-go/internal/ast/visitor.go:198 +0x44
github.com/microsoft/typescript-go/internal/ast.(*VariableStatement).VisitEachChild(0x4369001123c0, 0x43690068fd40)
        github.com/microsoft/typescript-go/internal/ast/ast_generated.go:1847 +0x44
github.com/microsoft/typescript-go/internal/ast.(*Node).VisitEachChild(...)
        github.com/microsoft/typescript-go/internal/ast/ast.go:197
github.com/microsoft/typescript-go/internal/ast.(*NodeVisitor).VisitEachChild(...)
        github.com/microsoft/typescript-go/internal/ast/visitor.go:191
github.com/microsoft/typescript-go/internal/transformers/estransforms.(*objectRestSpreadTransformer).visitVariableStatement(0x436901a9a780, 0x4369001123c0)
        github.com/microsoft/typescript-go/internal/transformers/estransforms/objectrestspread.go:416 +0x70
github.com/microsoft/typescript-go/internal/transformers/estransforms.(*objectRestSpreadTransformer).visit(0x436901a9a780, 0x4369001123c0)
        github.com/microsoft/typescript-go/internal/transformers/estransforms/objectrestspread.go:45 +0x370
github.com/microsoft/typescript-go/internal/ast.(*NodeVisitor).VisitSlice(0x43690068fd40, {0x4369001380e8, 0x1, 0x1})                                                                                         github.com/microsoft/typescript-go/internal/ast/visitor.go:148 +0xa0
github.com/microsoft/typescript-go/internal/ast.(*NodeVisitor).VisitNodes(0x43690068fd40, 0x4369001520a0)                                                                                                     github.com/microsoft/typescript-go/internal/ast/visitor.go:99 +0x3c
github.com/microsoft/typescript-go/internal/printer.(*EmitContext).VisitVariableEnvironment(0x43690071a0a0, 0x4369001520a0, 0x43690068fd40)
        github.com/microsoft/typescript-go/internal/printer/emitcontext.go:761 +0x17c
github.com/microsoft/typescript-go/internal/ast.(*NodeVisitor).visitTopLevelStatements(0x436901837b68?, 0x105179acc?)
        github.com/microsoft/typescript-go/internal/ast/visitor.go:255 +0x34                       github.com/microsoft/typescript-go/internal/ast.(*SourceFile).VisitEachChild(0x436900160008, 0x43690068fd40)                                                                                                  github.com/microsoft/typescript-go/internal/ast/ast.go:2599 +0x30
github.com/microsoft/typescript-go/internal/ast.(*Node).VisitEachChild(...)                                github.com/microsoft/typescript-go/internal/ast/ast.go:197
github.com/microsoft/typescript-go/internal/ast.(*NodeVisitor).VisitEachChild(...)
        github.com/microsoft/typescript-go/internal/ast/visitor.go:191                             github.com/microsoft/typescript-go/internal/transformers/estransforms.(*objectRestSpreadTransformer).visitSourceFile(0x436901a9a780, 0x105179acc?)
        github.com/microsoft/typescript-go/internal/transformers/estransforms/objectrestspread.go:72 +0x40
github.com/microsoft/typescript-go/internal/transformers/estransforms.(*objectRestSpreadTransformer).visit(0x436901a9a780, 0x436900160008)                                                                    github.com/microsoft/typescript-go/internal/transformers/estransforms/objectrestspread.go:31 +0x4f8
github.com/microsoft/typescript-go/internal/ast.(*NodeVisitor).VisitNode(0x0?, 0x1051edf20?)               github.com/microsoft/typescript-go/internal/ast/visitor.go:51 +0x30
github.com/microsoft/typescript-go/internal/ast.(*NodeVisitor).VisitSourceFile(...)
        github.com/microsoft/typescript-go/internal/ast/visitor.go:36
github.com/microsoft/typescript-go/internal/transformers.(*Transformer).TransformSourceFile(...)
        github.com/microsoft/typescript-go/internal/transformers/transformer.go:40
github.com/microsoft/typescript-go/internal/transformers.(*chainedTransformer).visit(0x436900696e80?, 0x436900160008?)                                                                                        github.com/microsoft/typescript-go/internal/transformers/chain.go:21 +0x8c
github.com/microsoft/typescript-go/internal/ast.(*NodeVisitor).VisitNode(0x43690071a0a0?, 0x10d033a00?)
        github.com/microsoft/typescript-go/internal/ast/visitor.go:51 +0x30                        github.com/microsoft/typescript-go/internal/ast.(*NodeVisitor).VisitSourceFile(...)
        github.com/microsoft/typescript-go/internal/ast/visitor.go:36                              github.com/microsoft/typescript-go/internal/transformers.(*Transformer).TransformSourceFile(...)
        github.com/microsoft/typescript-go/internal/transformers/transformer.go:40
github.com/microsoft/typescript-go/internal/compiler.(*emitter).runScriptTransformers(0x436900246e08?, 0x43690071a0a0, 0x436900160008)
        github.com/microsoft/typescript-go/internal/compiler/emitter.go:64 +0x178
                  github.com/microsoft/typescript-go/internal/compiler.(*emitter).emitJSFile(0x436900786e00, 0x436900160008, {0x4369000463c0, 0x2b}, {0x0, 0x0})
        github.com/microsoft/typescript-go/internal/compiler/emitter.go:192 +0x190                 github.com/microsoft/typescript-go/internal/compiler.(*emitter).emit(0x436900786e00)
        github.com/microsoft/typescript-go/internal/compiler/emitter.go:49 +0xec
github.com/microsoft/typescript-go/internal/compiler.(*Program).Emit.func2()                               github.com/microsoft/typescript-go/internal/compiler/program.go:1649 +0x194
github.com/microsoft/typescript-go/internal/core.(*parallelWorkGroup).Queue.func1()
        github.com/microsoft/typescript-go/internal/core/workgroup.go:40 +0x24                     sync.(*WaitGroup).Go.func1()
        sync/waitgroup.go:258 +0x48
created by sync.(*WaitGroup).Go in goroutine 1                                                             sync/waitgroup.go:238 +0x70
```

## Steps to reproduce

```ts
export var {...{}} = x;
```

```sh
tsgo --target es2017 bug.ts
```

--- a/internal/transformers/destructuring.go
+++ b/internal/transformers/destructuring.go
@@ -257,11 +257,17 @@ func (f *flattener) flattenDestructuringBinding(node *ast.Node, rval *ast.Node, 
 
 	if len(f.expressions) > 0 {
 		temp := f.tx.Factory().NewTempVariable()
-		f.tx.EmitContext().AddVariableDeclaration(temp)
-		last := &f.declarations[len(f.declarations)-1]
-		last.pendingExpressions = append(last.pendingExpressions, f.tx.Factory().NewAssignmentExpression(temp, last.value))
-		last.pendingExpressions = append(last.pendingExpressions, f.expressions...)
-		last.value = temp
+		if f.hoistTempVariables {
+			value := f.tx.Factory().InlineExpressions(f.expressions)
+			f.expressions = nil
+			f.emitBindingOrAssignment(f, temp, value, core.TextRange{}, nil)
+		} else {
+			f.tx.EmitContext().AddVariableDeclaration(temp)
+			last := &f.declarations[len(f.declarations)-1]
+			last.pendingExpressions = append(last.pendingExpressions, f.tx.Factory().NewAssignmentExpression(temp, last.value))
+			last.pendingExpressions = append(last.pendingExpressions, f.expressions...)
+			last.value = temp
+		}
 	}
 
 	decls := make([]*ast.Node, 0, len(f.declarations))

--- a/testdata/baselines/reference/compiler/destructuringEmptyBinding.errors.txt
+++ b/testdata/baselines/reference/compiler/destructuringEmptyBinding.errors.txt
@@ -1,0 +1,11 @@
+a.ts(1,16): error TS1003: Identifier expected.
+a.ts(1,23): error TS2304: Cannot find name 'x'.
+
+
+==== a.ts (2 errors) ====
+    export var {...{ }} = x;
+                   ~
+!!! error TS1003: Identifier expected.
+                          ~
+!!! error TS2304: Cannot find name 'x'.
+    

--- a/testdata/baselines/reference/compiler/destructuringEmptyBinding.js
+++ b/testdata/baselines/reference/compiler/destructuringEmptyBinding.js
@@ -1,0 +1,20 @@
+//// [tests/cases/compiler/destructuringEmptyBinding.ts] ////
+
+//// [a.ts]
+export var {...{ }} = x;
+
+
+//// [a.js]
+var __rest = (this && this.__rest) || function (s, e) {
+    var t = {};
+    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
+        t[p] = s[p];
+    if (s != null && typeof Object.getOwnPropertySymbols === "function")
+        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
+            if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
+                t[p[i]] = s[p[i]];
+        }
+    return t;
+};
+var _a;
+export var _b = _a = __rest(x, []);

--- a/testdata/baselines/reference/compiler/destructuringEmptyBinding.symbols
+++ b/testdata/baselines/reference/compiler/destructuringEmptyBinding.symbols
@@ -1,0 +1,6 @@
+//// [tests/cases/compiler/destructuringEmptyBinding.ts] ////
+
+=== a.ts ===
+
+export var {...{ }} = x;
+

--- a/testdata/baselines/reference/compiler/destructuringEmptyBinding.types
+++ b/testdata/baselines/reference/compiler/destructuringEmptyBinding.types
@@ -1,0 +1,7 @@
+//// [tests/cases/compiler/destructuringEmptyBinding.ts] ////
+
+=== a.ts ===
+export var {...{ }} = x;
+> : any
+>x : any
+

--- a/testdata/tests/cases/compiler/destructuringEmptyBinding.ts
+++ b/testdata/tests/cases/compiler/destructuringEmptyBinding.ts
@@ -1,0 +1,3 @@
+// @target: es2017
+// @filename: a.ts
+export var {...{ }} = x;

